components/raftstore/src/store/peer_storage.rs

```diff
@@ -1364,19 +1379,31 @@ where
         };
         // cleanup data before scheduling apply task
         if self.is_initialized() {
-            if let Err(e) = self.clear_extra_data(self.region()) {
+            if let Err(e) = self.clear_extra_data(self.get_region_id(), self.region(), &snap_region)
+            {
                 // No need panic here, when applying snapshot, the deletion will be tried
                 // again. But if the region range changes, like [a, c) -> [a, b) and [b, c),
                 // [b, c) will be kept in rocksdb until a covered snapshot is applied or
                 // store is restarted.
                 error!(
                     "failed to cleanup data, may leave some dirty data";
-                    "region_id" => self.region.get_id(),
+                    "region_id" => self.get_region_id(),
                     "peer_id" => self.peer_id,
                     "err" => ?e,
                 );
             }
         }
+        if let Some(ref destroyed_regions) = ctx.destroyed_regions {
+            for r in destroyed_regions {
+                if let Err(e) = self.clear_extra_data(r.get_id(), r, &snap_region) {
```

Comment:
What if `[a, b), [b, d)` tries to merge to `[a, c), [c, d)`? They seems to be racy on `[a, c)` and cause corruption.

Reply:
I don't get it. The regions in `destroyed_regions` have already written `tombstone`, so clear them is safe.

Reply:
The problem is `destroyed_regions` can be overlapped in different peers.

Reply:
For example
store1 (isolated): r1 [0, 100) r2 [100, 200) r3 [200,300)
1. r2 and r3 merge to r1
store2: r1 [0, 300)
2. r1 on store 2 sends snapshot to store 1.
3. `destroyed_regions` in r3 on store 1 is `[r2, r3]`
4. write r1, r2 to `tombstone` and r3 to `applying` atomically
5. clear r1, r2 data
6. remove r1,r2 and extend r3's range in `StoreMeta`